File: app/src/main/java/com/arman/queuetube/modules/playlists/GsonPlaylistHelper.java
package com.arman.queuetube.modules.playlists;

import android.content.Context;
import android.os.AsyncTask;

import com.arman.queuetube.activities.MainActivity;
import com.arman.queuetube.config.Constants;
import com.arman.queuetube.model.VideoData;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonPrimitive;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;

public class GsonPlaylistHelper {

    private static final JsonPrimitive PRIM_FAVORITES = new JsonPrimitive(Constants.Json.Playlist.FAVORITES);
    private static final JsonPrimitive PRIM_HISTORY = new JsonPrimitive(Constants.Json.Playlist.HISTORY);

    private static Context context;
    private static SavePlaylistsTask savePlaylistsTask;

    private static Gson gson;

    private static void executeSave(String string) {
        AsyncTask.Status status = savePlaylistsTask.getStatus();
        if (status != AsyncTask.Status.RUNNING) {
            if (status == AsyncTask.Status.FINISHED) {
                savePlaylistsTask = new SavePlaylistsTask((MainActivity) context);
            }
            savePlaylistsTask.execute(string);
        }
    }

    public static void doWrite(String string) throws IOException {
        FileOutputStream fos = context.openFileOutput(Constants.Json.STORAGE_FILE_NAME, Context.MODE_PRIVATE);
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(fos));
        bw.write(string);
        bw.flush();
        bw.close();
    }

    private static JsonObject doRead() throws IOException {
        FileInputStream fis = context.openFileInput(Constants.Json.STORAGE_FILE_NAME);
        BufferedReader br = new BufferedReader(new InputStreamReader(fis));
        StringBuilder sb = new StringBuilder();
        String line;
        while ((line = br.readLine()) != null) {
            sb.append(line);
        }
        JsonObject root = gson.fromJson(sb.toString(), JsonElement.class).getAsJsonObject();
        br.close();
        return root;
    }

    public static void onCreate(Context context) {
        GsonPlaylistHelper.context = context;
        savePlaylistsTask = new SavePlaylistsTask((MainActivity) context);
        gson = new GsonBuilder().setPrettyPrinting().create();
        boolean fileExists = doesFileExist();
        if (!fileExists) {
            JsonObject root = new JsonObject();
            JsonArray playlists = new JsonArray();
            playlists.add(newPlaylist(GsonPlaylistHelper.PRIM_HISTORY));
            playlists.add(newPlaylist(GsonPlaylistHelper.PRIM_FAVORITES));

            root.add(Constants.Json.Key.PLAYLISTS, playlists);
            executeSave(root.toString());
        }
    }

    public static JsonArray getFavorites() {
        return getPlaylist(read(), GsonPlaylistHelper.PRIM_FAVORITES);
    }

    public static JsonArray getHistory() {
        return getPlaylist(read(), GsonPlaylistHelper.PRIM_HISTORY);
    }

    public static boolean isFavorited(VideoData video) {
        JsonArray favorites = getFavorites();
        for (int i = 0; i < favorites.size(); i++) {
            JsonObject obj = favorites.get(i).getAsJsonObject();
            String id = obj.get(Constants.Json.Key.ID).getAsString();
            if (id.equals(video.getId())) {
                return true;
            }
        }
        return false;
    }

    private static JsonObject newPlaylist(JsonPrimitive name) {
        JsonObject obj = new JsonObject();
        obj.add(Constants.Json.Key.NAME, name);
        obj.add(Constants.Json.Key.PLAYLIST, new JsonArray());
        return obj;
    }

    private static JsonObject newPlaylist(String name) {
        JsonObject obj = new JsonObject();
        obj.addProperty(Constants.Json.Key.NAME, name);
        obj.add(Constants.Json.Key.PLAYLIST, new JsonArray());
        return obj;
    }

    private static JsonObject createVideo(VideoData video) {
        return gson.toJsonTree(video).getAsJsonObject();
    }

    public static JsonArray getPlaylists() {
        return getPlaylists(read());
    }

    private static JsonArray getPlaylists(JsonObject root) {
        return root.get(Constants.Json.Key.PLAYLISTS).getAsJsonArray();
    }

    public static JsonArray getUserPlaylists() {
        JsonObject root = read();
        JsonArray playlists = root.get(Constants.Json.Key.PLAYLISTS).getAsJsonArray();
        JsonArray userPlaylists = new JsonArray();
        for (JsonElement playlist : playlists) {
            JsonObject obj = playlist.getAsJsonObject();
            JsonPrimitive name = obj.getAsJsonPrimitive(Constants.Json.Key.NAME);
            if (!name.equals(PRIM_FAVORITES) && !name.equals(PRIM_HISTORY)) {
                userPlaylists.add(playlist);
            }
        }
        return userPlaylists;
    }

    public static JsonArray getPlaylist(JsonPrimitive name) {
        JsonObject root = read();
        return getPlaylist(root, name);
    }

    public static JsonArray getPlaylist(String name) {
        return getPlaylist(new JsonPrimitive(name));
    }

    private static JsonArray getPlaylist(JsonObject root, JsonPrimitive name) {
        JsonArray playlists = getPlaylists(root);
        assert playlists != null;
        for (int i = 0; i < playlists.size(); i++) {
            JsonObject obj = playlists.get(i).getAsJsonObject();
            if (obj.getAsJsonPrimitive(Constants.Json.Key.NAME).equals(name)) {
                return obj.get(Constants.Json.Key.PLAYLIST).getAsJsonArray();
            }
        }
        return null;
    }

    private static JsonArray getPlaylist(JsonObject root, String name) {
        return getPlaylist(root, new JsonPrimitive(name));
    }

    public static boolean writeNewIfNotFound(JsonPrimitive name) {
        JsonObject root = read();
        JsonArray playlists = getPlaylists(root);

        for (int i = 0; i < playlists.size(); i++) {
            JsonObject playlist = playlists.get(i).getAsJsonObject();
            if (playlist.getAsJsonPrimitive(Constants.Json.Key.NAME).equals(name)) {
                return false;
            }
        }

        JsonObject obj = newPlaylist(name);

        playlists.add(obj);
        executeSave(root.toString());
        return true;
    }

    public static boolean writeNewIfNotFound(String name) {
        return writeNewIfNotFound(new JsonPrimitive(name));
    }

    public static boolean writeNew(JsonPrimitive name) {
        JsonObject root = read();
        assert root != null;
        JsonArray playlists = getPlaylists(root);
        assert playlists != null;

        JsonObject obj = newPlaylist(name);

        playlists.add(obj);
        executeSave(root.toString());
        return true;
    }

    public static boolean writeNew(String name) {
        return writeNew(new JsonPrimitive(name));
    }

    private static boolean insert(JsonArray array, int pos, JsonObject obj) {
        array.add(new JsonObject());
        for (int i = array.size() - 1; i > pos; i--) {
            array.set(i, array.get(i - 1));
        }
        array.set(pos, obj);
        return true;
    }

    public static boolean reorder(JsonPrimitive name, int fromIndex, int toIndex) {
        JsonObject root = read();
        JsonArray playlist = getPlaylist(root, name);
        if (playlist != null) {
            JsonObject video = playlist.remove(fromIndex).getAsJsonObject();
            if (!insert(playlist, toIndex, video)) {
                return false;
            }
            executeSave(root.toString());
            return true;
        }
        return false;
    }

    public static boolean reorder(String name, int fromIndex, int toIndex) {
        return reorder(new JsonPrimitive(name), fromIndex, toIndex);
    }

    public static boolean reorder(JsonPrimitive name, VideoData video, int toIndex) {
        JsonObject root = read();
        JsonArray playlist = getPlaylist(root, name);
        if (playlist != null) {
            for (int i = 0; i < playlist.size(); i++) {
                JsonObject obj = playlist.get(i).getAsJsonObject();
                if (obj.get(Constants.Json.Key.ID).getAsString().equals(video.getId())) {
                    playlist.remove(i);
                    if (!insert(playlist, toIndex, obj)) {
                        return false;
                    }
                    executeSave(root.toString());
                    return true;
                }
            }
        }
        return false;
    }

    public static boolean reorder(String name, VideoData video, int toIndex) {
        return reorder(new JsonPrimitive(name), video, toIndex);
    }

    public static boolean removeFrom(JsonPrimitive name, VideoData video) {
        JsonObject root = read();
        JsonArray playlist = getPlaylist(root, name);
        if (playlist != null) {
            for (int i = 0; i < playlist.size(); i++) {
                JsonObject obj = playlist.get(i).getAsJsonObject();
                if (obj.get(Constants.Json.Key.ID).getAsString().equals(video.getId())) {
                    playlist.remove(i);
                    executeSave(root.toString());
                    return true;
                }
            }
        }
        return false;
    }

    public static boolean removeFrom(String name, VideoData video) {
        return removeFrom(new JsonPrimitive(name), video);
    }

    public static boolean writeToNew(JsonPrimitive name, VideoData video) {
        JsonObject root = read();
        JsonArray playlists = getPlaylists(root);

        JsonArray playlist = getPlaylist(root, name);
        if (playlist == null) {
            JsonObject obj = newPlaylist(name);
            playlists.add(obj);
            playlist = obj.getAsJsonArray(Constants.Json.Key.PLAYLIST);
        }
        if (playlist != null) {
            playlist.add(createVideo(video));
            executeSave(root.toString());
            return true;
        }
        return false;
    }

    public static boolean writeToNew(String name, VideoData video) {
        return writeToNew(new JsonPrimitive(name), video);
    }

    public static boolean writeToIfNotFound(JsonPrimitive name, VideoData video) {
        JsonObject root = read();
        JsonArray playlist = getPlaylist(root, name);
        if (playlist != null) {
            for (JsonElement jsonElement : playlist) {
                JsonObject obj = jsonElement.getAsJsonObject();
                if (obj.get(Constants.Json.Key.ID).getAsString().equals(video.getId())) {
                    return false;
                }
            }
            playlist.add(createVideo(video));
            executeSave(root.toString());
            return true;
        }
        return false;
    }

    public static boolean writeToIfNotFound(String name, VideoData video) {
        return writeToIfNotFound(new JsonPrimitive(name), video);
    }

    public static boolean writeToIfNotFound(JsonPrimitive name, VideoData video, int index) {
        JsonObject root = read();
        JsonArray playlist = getPlaylist(root, name);
        if (playlist != null) {
            for (JsonElement jsonElement : playlist) {
                JsonObject obj = jsonElement.getAsJsonObject();
                if (obj.get(Constants.Json.Key.ID).getAsString().equals(video.getId())) {
                    return false;
                }
            }
            if (!insert(playlist, index, createVideo(video))) {
                return false;
            }
            executeSave(root.toString());
            return true;
        }
        return false;
    }

    public static boolean writeToIfNotFound(String name, VideoData video, int index) {
        return writeToIfNotFound(new JsonPrimitive(name), video, index);
    }

    public static boolean writeToOrReorder(JsonPrimitive name, VideoData video, int index) {
        JsonObject root = read();
        JsonArray playlist = getPlaylist(root, name);
        if (playlist != null) {
            for (int i = 0; i < playlist.size(); i++) {
                JsonObject obj = playlist.get(i).getAsJsonObject();
                if (obj.get(Constants.Json.Key.ID).getAsString().equals(video.getId())) {
                    playlist.remove(i);
                    if (!insert(playlist, index, obj)) {
                        return false;
                    }
                    executeSave(root.toString());
                    return true;
                }
            }
        }
        return false;
    }

    public static boolean writeToOrReorder(String name, VideoData video, int index) {
        return writeToOrReorder(new JsonPrimitive(name), video, index);
    }

    public static boolean writeTo(JsonPrimitive name, VideoData video, int index) {
        JsonObject root = read();
        JsonArray playlist = getPlaylist(root, name);
        if (playlist != null) {
            if (!insert(playlist, index, createVideo(video))) {
                return false;
            }
            executeSave(root.toString());
            return true;
        }
        return false;
    }

    public static boolean writeTo(String name, VideoData video, int index) {
        return writeTo(new JsonPrimitive(name), video, index);
    }

    public static boolean writeTo(JsonPrimitive name, VideoData video) {
        JsonObject root = read();
        JsonArray playlist = getPlaylist(root, name);
        if (playlist != null) {
            playlist.add(createVideo(video));
            executeSave(root.toString());
            return true;
        }
        return false;
    }

    public static boolean writeTo(String name, VideoData video) {
        return writeTo(new JsonPrimitive(name), video);
    }

    private static void printInParts(String string) {
        int max = 1000;
        for (int i = 0; i <= string.length() / max; i++) {
            int start = i * max;
            int end = (i + 1) * max;
            end = end > string.length() ? string.length() : end;
            System.out.println(string.substring(start, end));
        }
    }

    public static JsonObject read() {
        try {
            return doRead();
        } catch (IOException e) {
            System.err.println(e.getMessage());
            return new JsonObject();
        }
    }

    private static boolean write(JsonObject obj) {
        String json = obj.toString();
        executeSave(json);
        return true;
    }

    private static boolean doesFileExist() {
        String path = context.getFilesDir().getAbsolutePath() + "/" + Constants.Json.STORAGE_FILE_NAME;
        File file = new File(path);
        return file.exists();
    }

    public static boolean remove(JsonPrimitive name) {
        JsonObject root = read();
        JsonArray playlists = getPlaylists(root);
        for (int i = 0; i < playlists.size(); i++) {
            JsonObject playlist = playlists.get(i).getAsJsonObject();
            if (playlist.getAsJsonPrimitive(Constants.Json.Key.NAME).equals(name)) {
                playlists.remove(i);
                executeSave(root.toString());
                return true;
            }
        }
        return false;
    }

    public static boolean remove(String name) {
        return remove(new JsonPrimitive(name));
    }

    public static boolean clear(JsonPrimitive name) {
        JsonObject root = read();
        JsonArray playlists = getPlaylists(root);
        for (int i = 0; i < playlists.size(); i++) {
            JsonObject playlist = playlists.get(i).getAsJsonObject();
            if (playlist.getAsJsonPrimitive(Constants.Json.Key.NAME).equals(name)) {
                playlists.remove(i);
                playlists.add(newPlaylist(name));
                executeSave(root.toString());
                return true;
            }
        }
        return false;
    }

    public static boolean clear(String name) {
        return clear(new JsonPrimitive(name));
    }

}

File: app/src/main/java/com/arman/queuetube/modules/playlists/json/GsonPlaylistHelper.kt
package com.arman.queuetube.modules.playlists.json

import android.annotation.SuppressLint
import android.content.Context
import android.os.AsyncTask
import com.arman.queuetube.config.Constants
import com.arman.queuetube.listeners.OnSaveFinishedListener
import com.arman.queuetube.listeners.OnTaskFinishedListener
import com.arman.queuetube.model.Video
import com.arman.queuetube.modules.BaseTask
import com.arman.queuetube.modules.InBackground
import com.google.gson.*
import java.io.*
import kotlin.math.min

@SuppressLint("StaticFieldLeak")
object GsonPlaylistHelper {

    private var context: Context? = null
    private val onSaveFinishedListeners: MutableList<OnSaveFinishedListener> = ArrayList()
    private var savePlaylistsTask: BaseTask<String, Unit>? = null

    private var gson: Gson? = null

    val favorites: JsonArray?
        get() = getPlaylist(Constants.Json.Playlist.FAVORITES)

    val history: JsonArray?
        get() = getPlaylist(Constants.Json.Playlist.HISTORY)

    val playlists: JsonArray?
        get() = getPlaylists(read())

    val userPlaylists: JsonArray
        get() {
            val root = read()
            val playlists = root.get(Constants.Json.Key.PLAYLISTS).asJsonArray
            val userPlaylists = JsonArray()
            for (playlist in playlists) {
                val obj = playlist.asJsonObject
                val name = obj.getAsJsonPrimitive(Constants.Json.Key.NAME).asString
                if (name != Constants.Json.Playlist.FAVORITES && name != Constants.Json.Playlist.HISTORY) {
                    userPlaylists.add(playlist)
                }
            }
            return userPlaylists
        }

    fun addOnSaveFinishedListener(onSaveFinishedListener: OnSaveFinishedListener) {
        this.onSaveFinishedListeners.add(onSaveFinishedListener)
    }

    fun removeOnSaveFinishedListener(onSaveFinishedListener: OnSaveFinishedListener) {
        this.onSaveFinishedListeners.remove(onSaveFinishedListener)
    }

    private fun doExecuteSave(vararg strings: String): Unit {
        try {
            doWrite(strings[0])
        } catch (e: IOException) {
        }
    }

    private fun executeSave(string: String = "", doInBackground: InBackground<String, Unit> = ::doExecuteSave) {
        if (savePlaylistsTask == null) {
            savePlaylistsTask = BaseTask(doInBackground)
            savePlaylistsTask?.onTaskFinishedListener = OnTaskFinishedListener {
                onSaveFinishedListeners.forEach { it.onSaveFinished() }
            }
        }
        val status = savePlaylistsTask!!.status
        if (status != AsyncTask.Status.RUNNING) {
            if (status == AsyncTask.Status.FINISHED) {
                savePlaylistsTask = BaseTask(doInBackground)
                savePlaylistsTask?.onTaskFinishedListener = OnTaskFinishedListener {
                    onSaveFinishedListeners.forEach { it.onSaveFinished() }
                }
            }
            savePlaylistsTask!!.execute(string)
        }
    }

    @Throws(IOException::class)
    fun doWrite(string: String) {
        val fos = context!!.openFileOutput(Constants.Json.STORAGE_FILE_NAME, Context.MODE_PRIVATE)
        val bw = BufferedWriter(OutputStreamWriter(fos))
        bw.write(string)
        bw.flush()
        bw.close()
    }

    @Throws(IOException::class)
    private fun doRead(): JsonObject {
        val fis = context!!.openFileInput(Constants.Json.STORAGE_FILE_NAME)
        val br = BufferedReader(InputStreamReader(fis))
        val sb = StringBuilder()
        var line: String?
        do {
            line = br.readLine()
            if (line == null) break
            sb.append(line)
        } while (true)
        val root = gson!!.fromJson(sb.toString(), JsonElement::class.java).asJsonObject
        br.close()
        return root
    }

    fun create(context: Context) {
        GsonPlaylistHelper.context = context
        gson = GsonBuilder().setPrettyPrinting().create()
        val fileExists = doesFileExist()
        if (!fileExists) {
            executeSave {
                val root = JsonObject()
                val playlists = JsonArray()
                playlists.add(newPlaylist(Constants.Json.Playlist.HISTORY))
                playlists.add(newPlaylist(Constants.Json.Playlist.FAVORITES))

                root.add(Constants.Json.Key.PLAYLISTS, playlists)
                doWrite(root.toString())
            }
        }
    }

    fun isFavorited(video: Video): Boolean {
        val favorites = favorites
        for (i in 0 until favorites!!.size()) {
            val obj = favorites.get(i).asJsonObject
            val id = obj.get(Constants.Json.Key.ID).asString
            if (id == video.id) {
                return true
            }
        }
        return false
    }

    fun playlistNames(array: JsonArray): MutableList<String> {
        val strings = java.util.ArrayList<String>()
        for (i in 0 until array.size()) {
            val playlist = array.get(i).asJsonObject
            strings.add(playlist.getAsJsonPrimitive(Constants.Json.Key.NAME).asString)
        }
        return strings
    }

    fun asPlaylist(array: JsonArray): MutableList<Video> {
        val videos = ArrayList<Video>()
        for (i in 0 until array.size()) {
            val video = array.get(i).asJsonObject
            val videoData = Video(video)
            videos.add(videoData)
        }
        return videos
    }

    fun asPlaylist(array: JsonArray, size: Int): MutableList<Video> {
        val videos = ArrayList<Video>()
        for (i in 0 until min(size, array.size())) {
            val video = array.get(i).asJsonObject
            val videoData = Video(video)
            videos.add(videoData)
        }
        return videos
    }

    private fun newPlaylist(name: String): JsonObject {
        val obj = JsonObject()
        obj.addProperty(Constants.Json.Key.NAME, name)
        obj.add(Constants.Json.Key.PLAYLIST, JsonArray())
        return obj
    }

    fun editName(fromName: String, toName: String): Boolean {
        val root = read()
        val playlists = getPlaylists(root)!!
        for (i in 0 until playlists.size()) {
            val obj = playlists.get(i).asJsonObject
            if (obj.getAsJsonPrimitive(Constants.Json.Key.NAME).asString == fromName) {
                executeSave {
                    obj.addProperty(Constants.Json.Key.NAME, toName)
                    doWrite(root.toString())
                }
                return true
            }
        }
        return false
    }

    private fun createVideo(video: Video): JsonObject {
        val obj = JsonObject()
        obj.addProperty(Constants.VideoData.ID, video.id)
        obj.addProperty(Constants.VideoData.TITLE, video.title)
        obj.addProperty(Constants.VideoData.CHANNEL_TITLE, video.channelTitle)
        obj.addProperty(Constants.VideoData.PUBLISHED_AT, video.publishedAt)
        obj.addProperty(Constants.VideoData.LIVE_BROADCAST_CONTENT, video.liveBroadcastContent)
        return obj
    }

    private fun getPlaylists(root: JsonObject): JsonArray? {
        return root.get(Constants.Json.Key.PLAYLISTS).asJsonArray
    }

    fun getPlaylist(name: String): JsonArray? {
        val root = read()
        return getPlaylist(root, name)
    }

    private fun getPlaylist(root: JsonObject, name: String): JsonArray? {
        val playlists = getPlaylists(root)!!
        for (i in 0 until playlists.size()) {
            val obj = playlists.get(i).asJsonObject
            if (obj.getAsJsonPrimitive(Constants.Json.Key.NAME).asString == name) {
                return obj.get(Constants.Json.Key.PLAYLIST).asJsonArray
            }
        }
        return null
    }

    fun writeNewIfNotFound(name: String): Boolean {
        val root = read()
        val playlists = getPlaylists(root)

        for (i in 0 until playlists!!.size()) {
            val playlist = playlists.get(i).asJsonObject
            if (playlist.getAsJsonPrimitive(Constants.Json.Key.NAME).asString == name) {
                return false
            }
        }

        executeSave {
            val obj = newPlaylist(name)
            playlists.add(obj)
            doWrite(root.toString())
        }

        return true
    }

    private fun insert(array: JsonArray, pos: Int, obj: JsonObject) {
        array.add(JsonObject())
        for (i in array.size() - 1 downTo pos + 1) {
            array.set(i, array.get(i - 1))
        }
        array.set(pos, obj)
    }

    fun reorder(name: String, fromIndex: Int, toIndex: Int): Boolean {
        val root = read()
        val playlist = getPlaylist(root, name)
        if (playlist != null) {
            executeSave {
                val video = playlist.remove(fromIndex).asJsonObject
                insert(playlist, toIndex, video)
                doWrite(root.toString())
            }
            return true
        }
        return false
    }

    fun reorder(name: String, video: Video, toIndex: Int): Boolean {
        val root = read()
        val playlist = getPlaylist(root, name)
        if (playlist != null) {
            for (i in 0 until playlist.size()) {
                val obj = playlist.get(i).asJsonObject
                if (obj.get(Constants.Json.Key.ID).asString == video.id) {
                    executeSave {
                        playlist.remove(i)
                        insert(playlist, toIndex, obj)
                        doWrite(root.toString())
                    }
                    return true
                }
            }
        }
        return false
    }

    fun removeFrom(name: String, video: Video): Boolean {
        val root = read()
        val playlist = getPlaylist(root, name)
        if (playlist != null) {
            for (i in 0 until playlist.size()) {
                val obj = playlist.get(i).asJsonObject
                if (obj.get(Constants.Json.Key.ID).asString == video.id) {
                    executeSave {
                        playlist.remove(i)
                        doWrite(root.toString())
                    }
                    return true
                }
            }
        }
        return false
    }

    fun writeToNew(name: String, video: Video): Boolean {
        val root = read()
        val playlists = getPlaylists(root)

        var playlist = getPlaylist(root, name)
        if (playlist == null) {
            val obj = newPlaylist(name)
            playlists!!.add(obj)
            playlist = obj.getAsJsonArray(Constants.Json.Key.PLAYLIST)
        }
        if (playlist != null) {
            executeSave {
                playlist.add(createVideo(video))
                doWrite(root.toString())
            }
            return true
        }
        return false
    }

    fun writeToIfNotFound(name: String, video: Video): Boolean {
        val root = read()
        val playlist = getPlaylist(root, name)
        if (playlist != null) {
            for (jsonElement in playlist) {
                val obj = jsonElement.asJsonObject
                if (obj.get(Constants.Json.Key.ID).asString == video.id) {
                    return false
                }
            }
            executeSave {
                playlist.add(createVideo(video))
                doWrite(root.toString())
            }
            return true
        }
        return false
    }

    fun writeToIfNotFound(name: String, video: Video, index: Int): Boolean {
        val root = read()
        val playlist = getPlaylist(root, name)
        if (playlist != null) {
            for (jsonElement in playlist) {
                val obj = jsonElement.asJsonObject
                if (obj.get(Constants.Json.Key.ID).asString == video.id) {
                    return false
                }
            }
            executeSave {
                insert(playlist, index, createVideo(video))
                doWrite(root.toString())
            }
            return true
        }
        return false
    }

    fun writeToOrReorder(name: String, video: Video, index: Int): Boolean {
        val root = read()
        val playlist = getPlaylist(root, name)
        if (playlist != null) {
            executeSave {
                for (i in 0 until playlist.size()) {
                    val obj = playlist.get(i).asJsonObject
                    if (obj.get(Constants.Json.Key.ID).asString == video.id) {
                        playlist.remove(i)
                        insert(playlist, index, obj)
                        doWrite(root.toString())
                        return@executeSave
                    }
                }
                insert(playlist, index, createVideo(video))
                doWrite(root.toString())
            }
            return true
        }
        return false
    }

    fun writeTo(name: String, video: Video, index: Int): Boolean {
        val root = read()
        val playlist = getPlaylist(root, name)
        if (playlist != null) {
            executeSave {
                insert(playlist, index, createVideo(video))
                doWrite(root.toString())
            }
            return true
        }
        return false
    }

    fun writeTo(name: String, video: Video): Boolean {
        val root = read()
        val playlist = getPlaylist(root, name)
        if (playlist != null) {
            executeSave {
                playlist.add(createVideo(video))
                doWrite(root.toString())
            }
            return true
        }
        return false
    }

    fun read(): JsonObject {
        return try {
            doRead()
        } catch (e: IOException) {
            JsonObject()
        }

    }

    private fun write(obj: JsonObject): Boolean {
        val json = obj.toString()
        executeSave(json)
        return true
    }

    private fun doesFileExist(): Boolean {
        val path = context!!.filesDir.absolutePath + "/" + Constants.Json.STORAGE_FILE_NAME
        val file = File(path)
        return file.exists()
    }

    fun remove(name: String): Boolean {
        val root = read()
        val playlists = getPlaylists(root)
        for (i in 0 until playlists!!.size()) {
            val playlist = playlists.get(i).asJsonObject
            if (playlist.getAsJsonPrimitive(Constants.Json.Key.NAME).asString == name) {
                executeSave {
                    playlists.remove(i)
                    doWrite(root.toString())
                }
                return true
            }
        }
        return false
    }

    fun clear(name: String): Boolean {
        val root = read()
        val playlists = getPlaylists(root)
        for (i in 0 until playlists!!.size()) {
            val playlist = playlists.get(i).asJsonObject
            if (playlist.getAsJsonPrimitive(Constants.Json.Key.NAME).asString == name) {
                executeSave {
                    playlists.remove(i)
                    insert(playlists, i, newPlaylist(name))
                    doWrite(root.toString())
                }
                return true
            }
        }
        return false
    }

}
